pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,298 @@
+type FlowSchema struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// MatchingPriority controls the ordering the flow-schemas in the cluster. Its value must be greater than zero and
+	// also unique. The primary sort is based on MatchingPriority, ordered highest to lowest (20 before 10).
+	MatchingPriority int32
+	// RequestPriorityRef can reference a RequestPriority in the cluster. If the reference cannot be resolved, the
+	// flow-schema will ignored and marked as invalid in its status.
+	RequestPriorityRef RequestPriorityRef
+	// Distinguisher defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	Distinguisher *FlowDistinguisher
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []RuleWithUser
+}
+
+// FlowDistinguisher specifies the configuration of the distinguisher.
+type FlowDistinguisher struct {
+	// Source defines the source attribute we adopt to compute the identifier for the flow.
+	Source FlowDistinguisherSource
+}
+
+// FlowDistinguisherSource defines the input source of distinguishing flows.
+type FlowDistinguisherSource string
+
+// RequestPriorityRef contains information that points to the request-priority being used.
+type RequestPriorityRef struct {
+	// APIGroup is the group for the resource being referenced.
+	APIGroup string
+	// Kind is the type of resource being referenced.
+	Kind string
+	// Name is the name of resource being referenced
+	Name string
+}
+
+// RuleWithUser is a tuple of UserRule and RequestScopeRule.
+type RuleWithUser struct {
+	// User is is the user that this rule cares about.
+	User UserRule
+	// Rule is embedded, it describes other criteria of the rule, like
+	// APIGroups, APIVersions, Resources, etc.
+	Rule RequestScopeRule
+}
+
+// RuleWithUser is a tuple of User and Rule.
+type UserRule struct {
```

Comment:
If you're including Groups and Users together, this should be Subject to be consistent with RBAC.

Reply:
now I copied the following definition from the rbac world. 

```go
type UserInfo struct {
	// The name that uniquely identifies this user among all active users.
	// +optional
	Username *string
	// A unique value that identifies this user across time. If this user is
	// deleted and another user by the same name is added, they will have
	// different UIDs.
	// +optional
	UID *string
	// The names of groups this user is a part of.
	// +optional
	Groups []string
	// Any additional information provided by the authenticator.
	// +optional
	Extra map[string]ExtraValue
}
```
what's more the user can unspecified/nil, by which the rule will be applied to all users. wdyt?